vowpalwabbit/automl.cc

```diff
@@ -0,0 +1,917 @@
+{
+  return challenger.lower_bound > champ.ips;
+}
+
+bool interaction_config_manager::worse(const exclusion_config& challenger, const exclusion_config& champ) const
+{
+  // Dummy return false to remove unused variable warning
+  return (champ.lower_bound > challenger.ips) ? false : false;
+}
+
+void interaction_config_manager::update_champ()
+{
+  // Update ips and lower bound for live configs
+  for (size_t stride = 0; stride < scores.size(); ++stride)
+  {
+    float ips = scores[stride].current_ips();
+    distributionally_robust::ScoredDual sd = scores[stride].chisq.recompute_duals();
+    float lower_bound = static_cast<float>(sd.first);
+    configs[scores[stride].config_index].ips = ips;
+    configs[scores[stride].config_index].lower_bound = lower_bound;
+  }
+  exclusion_config champ_config = configs[scores[current_champ].config_index];
+  bool champ_change = false;
+
+  // compare lowerbound of any challenger to the ips of the champ, and switch whenever when the LB beats the champ
+  for (size_t config_index = 0; config_index < configs.size(); ++config_index)
+  {
+    if (configs[config_index].state == New || configs[config_index].state == Removed ||
+        scores[current_champ].config_index == config_index)
+    { continue; }
+    // If challenger is better ('better function from Chacha')
+    if (better(configs[config_index], champ_config))
+    {
+      champ_change = true;
+      if (configs[config_index].state == Live)
+      {
+        // Update champ with stride of live config
+        for (size_t stride = 0; stride < scores.size(); ++stride)
+        {
+          if (scores[stride].config_index == config_index)
+          {
+            if (scores[stride].regular)
+            {
+              scores[stride].regular = false;
+              scores[current_champ].regular = true;
+            }
+            current_champ = stride;
+            break;
+          }
+        }
+        champ_config = configs[config_index];
+      }
+      else
+      {
+        float worst_ips = std::numeric_limits<float>::infinity();
+        size_t worst_stride = 0;
+        for (size_t stride = 0; stride < scores.size(); ++stride)
+        {
+          if (configs[scores[stride].config_index].ips < worst_ips)
+          {
+            worst_ips = configs[scores[stride].config_index].ips;
+            worst_stride = stride;
+          }
+        }
+        configs[scores[worst_stride].config_index].lease *= 2;
+        configs[scores[worst_stride].config_index].state = Inactive;
+        scores[worst_stride].reset_stats();
+        scores[worst_stride].config_index = config_index;
+        configs[scores[worst_stride].config_index].state = Live;
+        gen_quadratic_interactions(worst_stride);
+        if (scores[worst_stride].regular)
+        {
+          scores[worst_stride].regular = false;
+          scores[current_champ].regular = true;
+        }
+        current_champ = worst_stride;
```

Comment:
why does the current champ get set to the worst stride? seems counter intuitive

Reply:
This is a very awkward case which Chacha specifies but might not make sense in a VW context. In `update_champ()`, we compare the champ against all configs (live and non-live). This is the case where a non-live config overtakes the champ (with some lower_bound from a past run). Instead of replacing the champ with the non-live config that overtook it, we find the worst live config, and replace that instead. So `worst_stride` specifies the worst live config, we swap that out with the previously non-live config which is now champ, then we set `current_champ = worst_stride` since that's now the index with the champ.